File: max_heap.py
class MaxHeap:
    def __init__(self):
        self.values = []

    def insert(self, val):
        v = self.values
        v.append(val)

        child = len(v) - 1
        parent = (child - 1) // 2

        def check_pos(lst, par, ch):
            if par < 0 or (lst[par] >= lst[ch]): return

            lst[par], lst[ch] = lst[ch], lst[par]
            child = par
            parent = (child - 1) // 2

            check_pos(lst, parent, child)

        check_pos(v, parent, child)


    def remove_root(self):
        v = self.values
        end = len(v) - 1

        v[0], v[end] = v[end], v[0]
        max = v.pop()

        def getChild(lst, i):
            if len(lst) - 1 <  (i * 2) + 1: return None
            if len(lst) - 1 <  (i * 2) + 2: return (i * 2) + 1

            return (i * 2) + 1 if lst[(i * 2) + 1] > lst[(i * 2) + 2] else (i * 2) + 2

        root = 0
        child = getChild(v, root)

        def check_pos(lst, par, ch):
            if (not ch or lst[par] >= lst[ch]): return

            lst[par], lst[ch] = lst[ch], lst[par]
            parent = ch
            child = getChild(lst, parent)

            check_pos(lst, parent, child)


        check_pos(v, root, child)

        return max



heap = MaxHeap()
heap.insert(14)
heap.insert(22)
heap.insert(35)
heap.insert(40)
heap.insert(4)
heap.insert(18)
heap.insert(30)

File: maxHeap.js
class maxHeap {
  constructor() {
    this.values = [];
  }

  insert(val) {
    const v = this.values;
    v.push(val);
    let index = v.length - 1;
    let parent_index = ~~((index - 1) / 2);


    while (v[parent_index] < v[index]) {
      [v[parent_index], v[index]] = [v[index], v[parent_index]];
      index = parent_index;
      parent_index = ~~((index - 1) / 2);
    }

  }

  // Recursive solution

  insertNode(val) {
    const v = this.values;
    v.push(val);
    let index = v.length - 1;
    let parent_index = ~~((index - 1) / 2);

    function checkPosition(arr, c, p){
      if (arr[p] >= arr[c]) {
        return;
      }

      [arr[p], arr[c]] = [arr[c], arr[p]];
      let child = p;
      let parent = ~~((child - 1) / 2);

      checkPosition(arr, child, parent);
    }

    checkPosition(v, index, parent_index)
  }




  extract_max() {
    const v = this.values;

    [v[0], v[v.length - 1]] = [v[v.length - 1], v[0]];

    const max = v.pop();

    let root = 0;
    let child = (v[(root * 2) + 1] > v[(root * 2) + 2]) || !v[(root * 2) + 2] ?
        (root * 2) + 1 : (root * 2) + 2;

    if (child) {
      while(v[root] < v[child]) {
        [v[root], v[child]] = [v[child], v[root]];
        root = child;
        child = (v[(root * 2) + 1] > v[(root * 2) + 2]) || !v[(root * 2) + 2] ?
        (root * 2) + 1 : (root * 2) + 2;
      }
    }

    return max;
  }

  // Recursive solution

  extractMax(){
    const v = this.values;

    [v[0], v[v.length - 1]] = [v[v.length - 1], v[0]];

    const max = v.pop();

    let root = 0;
    let child = this.getChild(v, root);

    function checkPosition(arr, p, c, fn) {
      if (!arr[c] || arr[p] >= arr[c]) return;

      [arr[p], arr[c]] = [arr[c], arr[p]];

      let par = c;
      let ch = fn(v, par);

      checkPosition(v, par, ch, fn);
    }

    checkPosition(v, root, child, this.getChild);

    return max;
  }

  getChild(arr, i){
    if (arr[(i * 2) + 1]) {
     return (arr[(i * 2) + 1] > arr[(i * 2) + 2]) || !arr[(i * 2) + 2] ?
        (i * 2) + 1 : (i * 2) + 2;
    }
    return null;
  }
}

let heap = new maxHeap();
heap.values = [41, 39, 33, 18, 36, 12];
heap.insertNode(55)
heap.extractMax();
heap.extractMax();
heap.extractMax();
heap.insertNode(22);
heap.extractMax();

console.log("heap", heap);